File: ProbStat/LLN_CLT_SampleStats.R
# Preamble -------------

# Teaching program to show LLN and CLT applied to 
  # mean, variance, skewness, kurtosis, regression coefficients
  # individual and joint normality tests on simulated and financial data

# Skills covered:
  # simulating data and calculating moments
  # plotting histogram and fitting density
  # adding lines to plots
  # writing a loop to fill an empty list
  # run and summarize linear regression coefficients
  # use regression coefficients in later analysis

if(!require(pacman)) install.packages("pacman")
pacman::p_load(fBasics,quantmod,MASS)

# Show LLN and CLT of the first four moments of simulated data 
# do CLT for both Normal and non-Normal (Bernoulli) samples

# Define the population ---------------

# Set the seed
set.seed(12345)
popMean = 2.5
popSD = 0.5
popSkew = 0.0
popKurt = 0.0 # Subtract 3 to get the kurtosis associated with the sample data
X <- rnorm(1000000,mean = popMean, sd=popSD)
hist(X, breaks=100)
abline(v = popMean, lwd = 3, lty = 2)

# Law of Large Numbers (LLN) -----------------

# Calculate moments for increasingly large samples
mean(sample(X,10, replace = TRUE))
mean(sample(X,100, replace = TRUE))
mean(sample(X,1000, replace = TRUE))
mean(sample(X,10000, replace = TRUE))
sd(sample(X,10, replace = TRUE))
sd(sample(X,100, replace = TRUE))
sd(sample(X,1000, replace = TRUE))
sd(sample(X,10000, replace = TRUE))
skewness(sample(X,10, replace = TRUE))
skewness(sample(X,100, replace = TRUE))
skewness(sample(X,1000, replace = TRUE))
skewness(sample(X,10000, replace = TRUE))
kurtosis(sample(X,10, replace = TRUE),method = "moment")
kurtosis(sample(X,100, replace = TRUE),method = "moment")
kurtosis(sample(X,1000, replace = TRUE),method = "moment")
kurtosis(sample(X,10000, replace = TRUE),method = "moment")


# Central Limit Theorem -------------------

# Now for CLT. Take repeated samples from X all of the same size.
# 10,000 samples of size 10, compare to samples of size 100, 1000, 10000 etc.
# With many samples of increasingly large size... 
# ...the distribution of the sample statistics will converge to Normal.

## Take Samples ------------

realz = 10000 # number of samples or bootstrap realizations
N = 1000      # set to 10, 100, 1000, 10000 
mean_list <- list()
sd_list  <- list()
skew_list <- list()
kurt_list <- list()
for(i in 1:realz) {
  mean_list[[i]] <- mean(sample(X,N, replace = TRUE))
  sd_list[[i]] <- sd(sample(X,N, replace = TRUE))
  skew_list[[i]] <- skewness(sample(X,N, replace = TRUE))
  kurt_list[[i]] <- kurtosis(sample(X,N, replace = TRUE))
}

## "True" Distributions of Sample Statistics -----------

# compute population distribution for sample mean 
# standard error of the mean is (Std. Dev of sample/square root N)
# x = seq(0,5,0.001) says make x-axis from 0 to 5 in increments of 0.001. 
mu = dnorm(x=seq(0,5,0.001), mean=popMean, sd=popSD/sqrt(N))

# compute population distribution for stdev
# trick for inverse chi-squared since R doesn't have one by default
# integrate to get the CDF
s = seq(0,1,0.001)
sigmaCDF = sqrt(((N-1)*popSD*popSD)/qchisq(p=s, df=N-1))
# then differentiate to get the PDF
sigma = numeric(length(s)-1)
for (i in 2:length(s)){
  sigma[i] = (s[i]-s[i-1])/(sigmaCDF[i-1]-sigmaCDF[i])
}

# Skew and kurtosis are Normally distributed with sd = sqrt(6/N) and sd = sqrt(24/N), respectively.
g = dnorm(x=seq(-3,3,0.001), mean=popSkew, sd=sqrt(6/N)) # Skewness 
g2 = dnorm(x=seq(-3,3,0.001), mean=popKurt, sd=sqrt(24/N)) # Kurtosis 

## Distribution of Mean ----------------
hist(unlist(mean_list), breaks = 500, 
     xlab = "Mean of 10,000 samples from X of size 10", 
     main = "CLT of sample mean",
     cex.main = 0.8)
abline(v = popMean, lwd = 3, col = "white", lty = 2)
# Distribution of the population vs sample mean
plot(density(unlist(mean_list)),ylim=c(0,max(mu)), xlab="Mean", ylab="Density", main="Distribution of the Mean")
points(seq(0,5,0.001), mu, type='l', lwd = 2, col = "red")
abline(v = mean(unlist(mean_list)), lwd = 1, col = "black", lty = 2)
abline(v = popMean, lwd = 3, col = "red", lty = 2)
legend("topright", legend=c("bootstrap density", "theoretical density", "bootstrap mean value", "population value"), 
       col=c("black", "red", "black", "red"), lty=c(1,1,2,2), cex=0.8)

## Distribution of Std. Dev. --------------
hist(unlist(var_list), breaks = 500, 
     xlab = "Variance of 10,000 samples from X of size 10", 
     main = "CLT of sample variance",
     cex.main = 0.8)
abline(v = popSD^2, lwd = 3, col = "white", lty = 2)
# Distribution of the population vs. sample standard deviation
plot(density(unlist(sd_list)), ylim=c(0,max(sigma)), xlab="Standard Deviation", ylab="Density", main="Distribution of the Std. Dev")
points(sigmaCDF, sigma, type='l', lwd = 2, col = "red")
abline(v = mean(unlist(sd_list)), lwd = 1, col = "black", lty = 2)
abline(v = popSD, lwd = 3, col = "red", lty = 2)
legend("topright", legend=c("bootstrap density", "theoretical density", "bootstrap mean value", "population value"), 
       col=c("black", "red", "black", "red"), lty=c(1,1,2,2), cex=0.8)

## Distribution of Skewness ----------------
hist(unlist(skew_list), breaks = 500, 
     xlab = "Skewness of 10,000 samples from X of size 10", 
     main = "CLT of sample skewness",
     cex.main = 0.8)
abline(v = popSkew, lwd = 3, col = "white", lty = 2)
# Distribution of the population vs sample skew
plot(density(unlist(skew_list)), ylim=c(0,max(g)), xlab="Skewness", ylab="Density", main="Distribution of the Skewness")
points(seq(-3,3,0.001), g, type='l', lwd = 2, col = "red")
abline(v = mean(unlist(skew_list)), lwd = 1, col = "black", lty = 2)
abline(v = popSkew, lwd = 3, col = "red", lty = 2)
legend("topright", legend=c("bootstrap density", "theoretical density", "bootstrap mean value", "population value"), 
       col=c("black", "red", "black", "red"), lty=c(1,1,2,2), cex=0.8)

## Distribution of Kurtosis ---------------
hist(unlist(kurt_list), breaks = 500, 
     xlab = "Kurtosis of 10,000 samples from X of size 10", 
     main = "CLT of sample kurtosis",
     cex.main = 0.8)
abline(v = popKurt, lwd = 3, col = "white", lty = 2)
# Distribution of the population vs sample kurtosis
plot(density(unlist(kurt_list)), ylim=c(0,max(g2)), xlab="Kurtosis", ylab="Density", main="Distribution of the Kurtosis")
points(seq(-3,3,0.001), g2, type='l', lwd = 2, col = "red")
abline(v = mean(unlist(kurt_list)), lwd = 1, col = "black", lty = 2)
abline(v = popKurt, lwd = 3, col = "red", lty = 2)
legend("topright", legend=c("bootstrap density", "theoretical density", "bootstrap mean value", "population value"), 
       col=c("black", "red", "black", "red"), lty=c(1,1,2,2), cex=0.8)

## CLT for Non-Normal Distribution --------------

# Show that this works even if we weren't sampling from a Normal distribution
# sample from coin flips
population <- sample(c(0,1), 1000000, replace = TRUE)
hist(population, main = "Non-normal", cex.main = 0.8)
abline(v = mean(population), lwd = 3, lty = 3)

mean_list <- list()
for(i in 1:realz) {
  mean_list[[i]] <- mean(sample(population, N, replace = TRUE))
}
hist(unlist(mean_list), main = "Distribution of averages", cex.main = 0.8, xlab = "Average of samples with N=100")
abline(v = 0.5, lwd = 3)

# LLN and CLT for Regression Coefficients -----------------

## Create the population ---------------

eps <- rnorm(1000000, mean=0, sd=2)
beta0 = 0.5
beta1 = 0.8
Y <- beta0 + beta1*X + eps

summary(lm(Y~X))

## LLN for coefficients ------------------
Z = cbind(Y,X)
s10 <- Z[sample(nrow(Z),10,replace=TRUE),]
summary(lm(s10[,1]~s10[,2]))
s100 <- Z[sample(nrow(Z),100,replace=TRUE),]
summary(lm(s100[,1]~s100[,2]))
s1000 <- Z[sample(nrow(Z),1000,replace=TRUE),]
summary(lm(s1000[,1]~s1000[,2]))
s10000 <- Z[sample(nrow(Z),10000,replace=TRUE),]
summary(lm(s10000[,1]~s10000[,2]))

## CLT for coefficients ------------------
b0_list <- list()
b1_list <- list()
for(i in 1:realz) {
  sNx <- Z[sample(nrow(Z),N,replace=TRUE),]
  b0_list[[i]] <- lm(sNx[,1]~sNx[,2])$coefficients[1]
  b1_list[[i]] <- lm(sNx[,1]~sNx[,2])$coefficients[2]
}
hist(unlist(b0_list), main = "Distribution of intercepts", cex.main = 0.8, xlab = "Intercepts with N=100")
abline(v = beta0, lwd = 3)
hist(unlist(b1_list), main = "Distribution of slopes", cex.main = 0.8, xlab = "Slopes with N=100")
abline(v = beta1, lwd = 3)

File: 1_ProbStat/python_versions/LLN_CLT_SampleStats.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# =============================================================================
# LAW OF LARGE NUMBERS AND CENTRAL LIMIT THEOREM FOR SAMPLE STATISTICS
# =============================================================================

# Import required packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
from scipy.stats import norm, chi2
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm

# Set plotting style
plt.style.use('seaborn-v0_8-whitegrid')
plt.rcParams['figure.figsize'] = (12, 6)

# =============================================================================
# HELPER FUNCTIONS
# =============================================================================

def calculate_skewness(data):
    """
    Calculate skewness of a data array
    
    Parameters:
    -----------
    data : array-like
        Data array
        
    Returns:
    --------
    float
        Skewness
    """
    return stats.skew(data)

def calculate_kurtosis(data):
    """
    Calculate excess kurtosis of a data array
    
    Parameters:
    -----------
    data : array-like
        Data array
        
    Returns:
    --------
    float
        Excess kurtosis (normal = 0)
    """
    return stats.kurtosis(data, fisher=True)  # Fisher=True for excess kurtosis

def run_regression(X, y):
    """
    Run simple linear regression and return coefficients
    
    Parameters:
    -----------
    X : array-like
        Independent variable
    y : array-like
        Dependent variable
        
    Returns:
    --------
    tuple
        (intercept, slope)
    """
    # Add constant for intercept
    X_sm = sm.add_constant(X)
    
    # Fit regression model
    model = sm.OLS(y, X_sm)
    results = model.fit()
    
    # Extract coefficients
    intercept = results.params[0]
    slope = results.params[1]
    
    return intercept, slope

def plot_histogram_with_line(data, line_value, title, xlabel, bins=50):
    """
    Plot histogram with vertical line at specified value
    
    Parameters:
    -----------
    data : array-like
        Data to plot
    line_value : float
        Value to draw vertical line at
    title : str
        Plot title
    xlabel : str
        X-axis label
    bins : int
        Number of histogram bins
    """
    plt.figure(figsize=(10, 6))
    plt.hist(data, bins=bins, alpha=0.7, edgecolor='black')
    plt.axvline(x=line_value, color='red', linestyle='--', linewidth=2)
    plt.title(title, fontsize=14)
    plt.xlabel(xlabel, fontsize=12)
    plt.ylabel('Frequency', fontsize=12)
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.show()

def plot_density_comparison(empirical_data, theoretical_x, theoretical_y, 
                           title, xlabel, pop_value, sample_mean=None):
    """
    Plot empirical density vs theoretical density
    
    Parameters:
    -----------
    empirical_data : array-like
        Empirical data
    theoretical_x : array-like
        X values for theoretical density
    theoretical_y : array-like
        Y values for theoretical density
    title : str
        Plot title
    xlabel : str
        X-axis label
    pop_value : float
        Population parameter value
    sample_mean : float, optional
        Sample mean value
    """
    plt.figure(figsize=(10, 6))
    
    # Plot empirical density
    sns.kdeplot(empirical_data, label='Bootstrap Density')
    
    # Plot theoretical density
    plt.plot(theoretical_x, theoretical_y, 'r-', linewidth=2, label='Theoretical Density')
    
    # Add vertical lines
    if sample_mean is not None:
        plt.axvline(x=sample_mean, color='black', linestyle='--', linewidth=1, label='Bootstrap Mean Value')
    plt.axvline(x=pop_value, color='red', linestyle='--', linewidth=2, label='Population Value')
    
    plt.title(title, fontsize=14)
    plt.xlabel(xlabel, fontsize=12)
    plt.ylabel('Density', fontsize=12)
    plt.legend()
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.show()

# =============================================================================
# MAIN FUNCTIONS
# =============================================================================

def demonstrate_lln_sample_moments(X, pop_mean, pop_sd, pop_skew, pop_kurt):
    """
    Demonstrate Law of Large Numbers for sample moments
    
    Parameters:
    -----------
    X : array-like
        Population data
    pop_mean : float
        Population mean
    pop_sd : float
        Population standard deviation
    pop_skew : float
        Population skewness
    pop_kurt : float
        Population excess kurtosis
    """
    print("\n" + "=" * 70)
    print("DEMONSTRATION OF LAW OF LARGE NUMBERS (LLN)")
    print("=" * 70)
    
    # Sample sizes to demonstrate
    sample_sizes = [10, 100, 1000, 10000]
    
    # Create empty lists to store results
    mean_results = []
    sd_results = []
    skew_results = []
    kurt_results = []
    
    # Calculate moments for each sample size
    for size in sample_sizes:
        sample = np.random.choice(X, size=size, replace=True)
        mean_results.append(np.mean(sample))
        sd_results.append(np.std(sample, ddof=1))
        skew_results.append(calculate_skewness(sample))
        kurt_results.append(calculate_kurtosis(sample))
    
    # Create a results DataFrame
    results = pd.DataFrame({
        'Sample Size': sample_sizes,
        'Mean': mean_results,
        'Std Dev': sd_results,
        'Skewness': skew_results,
        'Kurtosis': kurt_results
    })
    
    # Add population values for comparison
    pop_values = pd.DataFrame({
        'Sample Size': ['Population'],
        'Mean': [pop_mean],
        'Std Dev': [pop_sd],
        'Skewness': [pop_skew],
        'Kurtosis': [pop_kurt]
    })
    
    results = pd.concat([results, pop_values])
    
    print("\nSample Moments as Sample Size Increases:")
    print(results.to_string(index=False))
    
    # Plot convergence
    plt.figure(figsize=(15, 10))
    
    plt.subplot(2, 2, 1)
    plt.plot(sample_sizes, mean_results, 'o-', linewidth=2)
    plt.axhline(y=pop_mean, color='r', linestyle='--', linewidth=2)
    plt.xscale('log')
    plt.title('Mean Convergence', fontsize=14)
    plt.xlabel('Sample Size (log scale)', fontsize=12)
    plt.ylabel('Sample Mean', fontsize=12)
    
    plt.subplot(2, 2, 2)
    plt.plot(sample_sizes, sd_results, 'o-', linewidth=2)
    plt.axhline(y=pop_sd, color='r', linestyle='--', linewidth=2)
    plt.xscale('log')
    plt.title('Standard Deviation Convergence', fontsize=14)
    plt.xlabel('Sample Size (log scale)', fontsize=12)
    plt.ylabel('Sample Std Dev', fontsize=12)
    
    plt.subplot(2, 2, 3)
    plt.plot(sample_sizes, skew_results, 'o-', linewidth=2)
    plt.axhline(y=pop_skew, color='r', linestyle='--', linewidth=2)
    plt.xscale('log')
    plt.title('Skewness Convergence', fontsize=14)
    plt.xlabel('Sample Size (log scale)', fontsize=12)
    plt.ylabel('Sample Skewness', fontsize=12)
    
    plt.subplot(2, 2, 4)
    plt.plot(sample_sizes, kurt_results, 'o-', linewidth=2)
    plt.axhline(y=pop_kurt, color='r', linestyle='--', linewidth=2)
    plt.xscale('log')
    plt.title('Kurtosis Convergence', fontsize=14)
    plt.xlabel('Sample Size (log scale)', fontsize=12)
    plt.ylabel('Sample Kurtosis', fontsize=12)
    
    plt.tight_layout()
    plt.show()

def demonstrate_clt_sample_moments(X, pop_mean, pop_sd, pop_skew, pop_kurt, realz=10000, N=1000):
    """
    Demonstrate Central Limit Theorem for sample moments
    
    Parameters:
    -----------
    X : array-like
        Population data
    pop_mean : float
        Population mean
    pop_sd : float
        Population standard deviation
    pop_skew : float
        Population skewness
    pop_kurt : float
        Population excess kurtosis
    realz : int
        Number of bootstrap realizations
    N : int
        Sample size for each realization
    """
    print("\n" + "=" * 70)
    print(f"DEMONSTRATION OF CENTRAL LIMIT THEOREM (CLT) WITH N={N}")
    print("=" * 70)
    
    # Initialize empty lists for storing results
    mean_list = []
    sd_list = []
    skew_list = []
    kurt_list = []
    
    # Take repeated samples and calculate statistics
    print(f"\nTaking {realz} samples of size {N}...")
    for _ in range(realz):
        sample = np.random.choice(X, size=N, replace=True)
        mean_list.append(np.mean(sample))
        sd_list.append(np.std(sample, ddof=1))
        skew_list.append(calculate_skewness(sample))
        kurt_list.append(calculate_kurtosis(sample))
    
    # Convert lists to arrays
    mean_array = np.array(mean_list)
    sd_array = np.array(sd_list)
    skew_array = np.array(skew_list)
    kurt_array = np.array(kurt_list)
    
    # Calculate theoretical distributions
    # For mean: Normal with mean=pop_mean and sd=pop_sd/sqrt(N)
    x_mean = np.linspace(pop_mean - 4*pop_sd/np.sqrt(N), 
                         pop_mean + 4*pop_sd/np.sqrt(N), 1000)
    y_mean = norm.pdf(x_mean, loc=pop_mean, scale=pop_sd/np.sqrt(N))
    
    # For standard deviation: derived from chi-square distribution
    x_sd = np.linspace(max(0.001, pop_sd - 4*pop_sd/np.sqrt(2*N)), 
                       pop_sd + 4*pop_sd/np.sqrt(2*N), 1000)
    # Approximation for large N - standard deviation of sample sd
    sd_of_sd = pop_sd / np.sqrt(2*N)
    y_sd = norm.pdf(x_sd, loc=pop_sd, scale=sd_of_sd)
    
    # For skewness: Normal with mean=pop_skew and sd=sqrt(6/N)
    x_skew = np.linspace(pop_skew - 4*np.sqrt(6/N), 
                         pop_skew + 4*np.sqrt(6/N), 1000)
    y_skew = norm.pdf(x_skew, loc=pop_skew, scale=np.sqrt(6/N))
    
    # For kurtosis: Normal with mean=pop_kurt and sd=sqrt(24/N)
    x_kurt = np.linspace(pop_kurt - 4*np.sqrt(24/N), 
                         pop_kurt + 4*np.sqrt(24/N), 1000)
    y_kurt = norm.pdf(x_kurt, loc=pop_kurt, scale=np.sqrt(24/N))
    
    # Display summary statistics
    print("\nSummary of Bootstrap Distributions:")
    print(f"Mean: {np.mean(mean_array):.6f} (Theoretical: {pop_mean})")
    print(f"Std Dev: {np.mean(sd_array):.6f} (Theoretical: {pop_sd})")
    print(f"Skewness: {np.mean(skew_array):.6f} (Theoretical: {pop_skew})")
    print(f"Kurtosis: {np.mean(kurt_array):.6f} (Theoretical: {pop_kurt})")
    
    # Plot histograms
    plot_histogram_with_line(
        mean_array, pop_mean, 
        f"CLT of Sample Mean (N={N})", 
        f"Mean of {realz} samples from X of size {N}"
    )
    
    plot_histogram_with_line(
        sd_array, pop_sd, 
        f"CLT of Sample Standard Deviation (N={N})", 
        f"Std Dev of {realz} samples from X of size {N}"
    )
    
    plot_histogram_with_line(
        skew_array, pop_skew, 
        f"CLT of Sample Skewness (N={N})", 
        f"Skewness of {realz} samples from X of size {N}"
    )
    
    plot_histogram_with_line(
        kurt_array, pop_kurt, 
        f"CLT of Sample Kurtosis (N={N})", 
        f"Kurtosis of {realz} samples from X of size {N}"
    )
    
    # Plot density comparisons
    plot_density_comparison(
        mean_array, x_mean, y_mean, 
        "Distribution of the Mean", "Mean", 
        pop_mean, np.mean(mean_array)
    )
    
    plot_density_comparison(
        sd_array, x_sd, y_sd, 
        "Distribution of the Standard Deviation", "Standard Deviation", 
        pop_sd, np.mean(sd_array)
    )
    
    plot_density_comparison(
        skew_array, x_skew, y_skew, 
        "Distribution of the Skewness", "Skewness", 
        pop_skew, np.mean(skew_array)
    )
    
    plot_density_comparison(
        kurt_array, x_kurt, y_kurt, 
        "Distribution of the Kurtosis", "Kurtosis", 
        pop_kurt, np.mean(kurt_array)
    )

def demonstrate_clt_non_normal(realz=10000, N=1000):
    """
    Demonstrate CLT for non-normal distribution (Bernoulli)
    
    Parameters:
    -----------
    realz : int
        Number of bootstrap realizations
    N : int
        Sample size for each realization
    """
    print("\n" + "=" * 70)
    print(f"DEMONSTRATION OF CLT FOR NON-NORMAL DISTRIBUTION WITH N={N}")
    print("=" * 70)
    
    # Create Bernoulli population (coin flips)
    population = np.random.choice([0, 1], size=1000000, replace=True)
    pop_mean = np.mean(population)
    
    # Plot population distribution
    plt.figure(figsize=(10, 6))
    plt.hist(population, bins=3, alpha=0.7, edgecolor='black')
    plt.axvline(x=pop_mean, color='red', linestyle='--', linewidth=2)
    plt.title("Non-Normal Population Distribution (Bernoulli)", fontsize=14)
    plt.xlabel("Value", fontsize=12)
    plt.ylabel("Frequency", fontsize=12)
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.show()
    
    # Calculate sample means
    mean_list = []
    for _ in range(realz):
        sample = np.random.choice(population, size=N, replace=True)
        mean_list.append(np.mean(sample))
    
    # Convert to array
    mean_array = np.array(mean_list)
    
    # Theoretical distribution for sample mean
    x_mean = np.linspace(pop_mean - 4*np.sqrt(0.25/N), 
                         pop_mean + 4*np.sqrt(0.25/N), 1000)
    y_mean = norm.pdf(x_mean, loc=pop_mean, scale=np.sqrt(0.25/N))
    
    # Plot histogram
    plot_histogram_with_line(
        mean_array, pop_mean, 
        f"CLT of Sample Mean for Bernoulli Distribution (N={N})", 
        f"Mean of {realz} samples from Bernoulli of size {N}"
    )
    
    # Plot density comparison
    plot_density_comparison(
        mean_array, x_mean, y_mean, 
        "Distribution of the Mean (Bernoulli)", "Mean", 
        pop_mean, np.mean(mean_array)
    )

def demonstrate_regression_lln_clt(X, realz=10000, N=1000):
    """
    Demonstrate LLN and CLT for regression coefficients
    
    Parameters:
    -----------
    X : array-like
        Independent variable values
    realz : int
        Number of bootstrap realizations
    N : int
        Sample size for each realization
    """
    print("\n" + "=" * 70)
    print("DEMONSTRATION OF LLN AND CLT FOR REGRESSION COEFFICIENTS")
    print("=" * 70)
    
    # Set true parameter values
    beta0 = 0.5
    beta1 = 0.8
    
    # Generate error term
    eps = np.random.normal(0, 2, size=len(X))
    
    # Generate dependent variable
    Y = beta0 + beta1 * X + eps
    
    # Combine X and Y into a single array for sampling
    Z = np.column_stack((Y, X))
    
    # Run regression on population
    intercept, slope = run_regression(X, Y)
    print("\nRegression on Population:")
    print(f"Intercept: {intercept:.6f} (True: {beta0})")
    print(f"Slope: {slope:.6f} (True: {beta1})")
    
    # Demonstrate LLN for regression coefficients
    print("\nLLN for Regression Coefficients:")
    sample_sizes = [10, 100, 1000, 10000]
    
    for size in sample_sizes:
        # Sample from population
        indices = np.random.choice(len(Z), size=size, replace=True)
        sample_z = Z[indices]
        sample_y = sample_z[:, 0]
        sample_x = sample_z[:, 1]
        
        # Run regression
        intercept, slope = run_regression(sample_x, sample_y)
        print(f"Sample Size {size}: Intercept = {intercept:.6f}, Slope = {slope:.6f}")
    
    # Demonstrate CLT for regression coefficients
    print("\nCLT for Regression Coefficients:")
    b0_list = []
    b1_list = []
    
    for _ in range(realz):
        # Sample from population
        indices = np.random.choice(len(Z), size=N, replace=True)
        sample_z = Z[indices]
        sample_y = sample_z[:, 0]
        sample_x = sample_z[:, 1]
        
        # Run regression
        intercept, slope = run_regression(sample_x, sample_y)
        b0_list.append(intercept)
        b1_list.append(slope)
    
    # Convert to arrays
    b0_array = np.array(b0_list)
    b1_array = np.array(b1_list)
    
    # Plot histograms
    plot_histogram_with_line(
        b0_array, beta0, 
        f"Distribution of Regression Intercepts (N={N})", 
        f"Intercepts from {realz} samples of size {N}"
    )
    
    plot_histogram_with_line(
        b1_array, beta1, 
        f"Distribution of Regression Slopes (N={N})", 
        f"Slopes from {realz} samples of size {N}"
    )

# =============================================================================
# MAIN FUNCTION
# =============================================================================

if __name__ == "__main__":
    print("\nPython version of LLN_CLT_SampleStats.R")
    print("=" * 70)
    
    # Set random seed for reproducibility
    np.random.seed(12345)
    
    # Define population parameters
    pop_mean = 2.5
    pop_sd = 0.5
    pop_skew = 0.0
    pop_kurt = 0.0  # Excess kurtosis (normal = 0)
    
    # Generate population data
    X = np.random.normal(loc=pop_mean, scale=pop_sd, size=1000000)
    
    # Plot population distribution
    plt.figure(figsize=(10, 6))
    plt.hist(X, bins=100, alpha=0.7, edgecolor='black')
    plt.axvline(x=pop_mean, color='red', linestyle='--', linewidth=2)
    plt.title(f"Population Distribution (N={len(X):,})", fontsize=14)
    plt.xlabel("Value", fontsize=12)
    plt.ylabel("Frequency", fontsize=12)
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.show()
    
    # Run demonstrations
    demonstrate_lln_sample_moments(X, pop_mean, pop_sd, pop_skew, pop_kurt)
    
    # For CLT, can adjust N to 10, 100, 1000, 10000 to see the effect of sample size
    demonstrate_clt_sample_moments(X, pop_mean, pop_sd, pop_skew, pop_kurt, 
                                  realz=10000, N=1000)
    
    # Demonstrate CLT for non-normal distribution
    demonstrate_clt_non_normal(realz=10000, N=1000)
    
    # Demonstrate LLN and CLT for regression coefficients
    demonstrate_regression_lln_clt(X, realz=10000, N=1000)